datacube_apps/stacker/fixer.py

```diff
@@ -0,0 +1,273 @@
+"""
+Create time-stacked NetCDF files
+
+"""
+from __future__ import absolute_import, print_function
+
+import copy
+import datetime
+import itertools
+import logging
+import os
+import re
+import socket
+from functools import partial
+from collections import Counter
+
+import click
+import dask.array as da
+import dask
+from dateutil import tz
+from pathlib import Path
+import pandas as pd
+import xarray as xr
+
+import datacube
+from datacube.model import Dataset
+from datacube.model.utils import xr_apply, datasets_to_doc
+from datacube.storage import netcdf_writer
+from datacube.storage.storage import create_netcdf_storage_unit
+from datacube.ui import task_app
+from datacube.ui.click import to_pathlib
+
+
+_LOG = logging.getLogger(__name__)
+
+
+APP_NAME = 'datacube-fixer'
+
+
+def get_filename(config, cell_index, year):
+    file_path_template = str(Path(config['location'], config['file_path_template']))
+    return file_path_template.format(tile_index=cell_index, start_time=year, version=config['taskfile_version'])
+
+
+def get_temp_file(final_output_path):
+    """
+    Get a temp file path
+    Changes "/path/file.nc" to "/path/.tmp/file.nc.host.pid.tmp"
+    :param Path final_output_path:
+    :return: Path to temporarily write output
+    :rtype: Path
+    """
+
+    tmp_folder = final_output_path.parent / '.tmp'
+    id_file = '{host}.{pid}'.format(host=socket.gethostname(), pid=os.getpid())
+    tmp_path = (tmp_folder / final_output_path.stem).with_suffix(final_output_path.suffix + id_file + '.tmp')
+    try:
+        tmp_folder.mkdir(parents=True)
+    except OSError:
+        pass
+    if tmp_path.exists():
+        tmp_path.unlink()
+    return tmp_path
+
+
+FIND_TIME_RE = re.compile(r'.+_(?P<start_time>\d+)(?:_v\d+)?\.nc\Z')
+
+
+def get_single_dataset_paths(cell):
+    cnt = Counter(ds.local_path for ds in itertools.chain(*cell.sources.values))
+    files_to_fix = [local_path for local_path, count in cnt.items()
+                    if count == 1 and FIND_TIME_RE.search(str(local_path)).groups()]
+    return files_to_fix
+
+
+def make_fixer_tasks(index, config, **kwargs):
+    # Find tiles with bad metadata
+    # In this case, they will be single timeslices
+    # Typically for particular years: the first and last for each product
+    # But for 2016, nothing will be stacked
+    # So we could do it based on NetCDF properties, and for all single-dataset locations
+    # Tile -> locations -> unique locations -> tile
+    # Just regex the filenames
+
+    query = {kw: arg for kw, arg in kwargs.items() if kw in ['cell_index'] and arg is not None}
+    gw = datacube.api.GridWorkflow(index=index, product=config['product'].name)
+
+    time_query_list = task_app.year_splitter(*kwargs['time']) if 'time' in kwargs else [None]
+
+    for time_query in time_query_list:
+        cells = gw.list_cells(product=config['product'].name, time=time_query, **query)
+        for cell_index, cell in cells.items():
```

Comment:
It is doing too much. It is also mishandling time, which I'll fix.
Ideally `fixer` would become a special case of `stacker`, but right now fixer has a hardcoded regex on the filename to get the time value to determine if a file is stacked or not.

I'd like to keep `task_app` unaware of GridWorkflow. (`task_app` should probably have `index` removed from it too)